### PR TITLE
Feature Reset Target Attributes

### PR DESCRIPTION
**Pull request by stefbehl**
Introduces an "update mode" parameter for all update operations on target attributes in DDI and DMF.

The update mode can have the following values:
- MERGE: Merges the attributes that are part of the request into the existing attributes. This is the default.
- REPLACE: Clears all existing attributes and replaces them with the attributes specified in the request. This is the update mode that can be used for resetting the target attributes.
- REMOVE: Removes all attributes that are specified in the request.

Affected API endoints:
- DDI: PUT /controller/v1/{targetid}/configData
- DMF: UpdateTargetAttributes AMQP message

In both cases, the message payload can now contain an (optional) JSON property "mode" which specifies the update mode that should be applied during the update operation.

### CODE REVIEW

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/UpdateMode.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.repository;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+public enum UpdateMode {
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/UpdateMode.java]: Add JavaDoc

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java [hunk 2]
```diff
@@ -666,18 +667,38 @@ private void handleFinishedAndStoreInTargetStatus(final JpaAction action) {
...
-                    quotaManagement.getMaxAttributeEntriesPerTarget());
+        switch (mode) {
+        case REMOVE:
+            // remove the addressed attributes
+            data.keySet().forEach(controllerAttributes::remove);
+            break;
+        case REPLACE:
+            // clear the attributes before adding the new attributes
+            controllerAttributes.clear();
+            // redundant code to satisfy Sonar
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: remove sonar comment

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java [hunk 3]
```diff
@@ -666,18 +667,38 @@ private void handleFinishedAndStoreInTargetStatus(final JpaAction action) {
...
+            controllerAttributes.putAll(data);
+            target.setRequestControllerAttributes(false);
+            break;
+        default:
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: better have a separate case for MERGE and throw an exception in default case

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ControllerManagementTest.java [hunk 4]
```diff
@@ -678,6 +680,67 @@ private void updateAttributeAndVerify(final String controllerId) {
                 .isEqualTo(testData);
     }
 
+    @Test
```
Comment (stefbehl) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/ControllerManagementTest.java]: keep one test method but move different test cases (for different update modes) into separate methods which are annotated with Step annotation

File: hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/AmqpMessageHandlerServiceIntegrationTest.java [hunk 5]
```diff
@@ -634,23 +637,70 @@ public void canceledRejectedActionStatus() {
     }
 
     @Test
-    @Description("Verify that sending an update controller attribute message to an existing target works.")
+    @Description("Verify that sending an update controller attribute message to an existing target works. Verify that different update modes (merge, replace, remove) can be used.")
     @ExpectEvents({ @Expect(type = TargetCreatedEvent.class, count = 1),
-            @Expect(type = TargetUpdatedEvent.class, count = 1), @Expect(type = TargetPollEvent.class, count = 1) })
+            @Expect(type = TargetUpdatedEvent.class, count = 4), @Expect(type = TargetPollEvent.class, count = 1) })
```
Comment (stefbehl) [inline hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/integration/AmqpMessageHandlerServiceIntegrationTest.java]: move into separate methods annotated with Step annotation

File: hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerServiceTest.java [hunk 6]
```diff
@@ -202,6 +207,50 @@ public void updateAttributes() {
 
     }
 
+    @Test
```
Comment (stefbehl) [inline hawkbit-dmf/hawkbit-dmf-amqp/src/test/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerServiceTest.java]: use separate methods annotated with Step annotation

File: hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java [hunk 7]
```diff
@@ -364,4 +367,17 @@ private Action checkActionExist(final Message message, final DmfActionUpdateStat
...
     }
+
+    /**
+     * Retrieve the update mode from the given update message.
+     */
+    private static UpdateMode getUpdateMode(final DmfAttributeUpdate update) {
+        final DmfUpdateMode mode = update.getMode();
+        if (mode != null) {
+            return UpdateMode.valueOf(mode.name());
+        } else {
```
Comment (stefbehl) [inline hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java]: remove else

File: hawkbit-ddi-resource/src/main/java/org/eclipse/hawkbit/ddi/rest/resource/DdiRootController.java [hunk 8]
```diff
@@ -426,10 +428,14 @@ private Status handleClosedCase(final DdiActionFeedback feedback, final String c
 
     @Override
     public ResponseEntity<Void> putConfigData(@Valid @RequestBody final DdiConfigData configData,
-            @PathVariable("tenant") final String tenant, @PathVariable("controllerId") final String controllerId) {
-        controllerManagement.updateControllerAttributes(controllerId, configData.getData());
+            @PathVariable("tenant") final String tenant, @PathVariable("controllerId") final String controllerId,
+            @RequestParam(value = "mode", required = false, defaultValue = DdiRestConstants.DEFAULT_UPDATE_MODE) final String mode) {
 
-        return ResponseEntity.ok().build();
+        // deserialize the update mode, then invoke the controller mgmt
```
Comment (stefbehl) [inline hawkbit-ddi-resource/src/main/java/org/eclipse/hawkbit/ddi/rest/resource/DdiRootController.java]: retrieve mode from payload

File: hawkbit-ddi-resource/src/main/java/org/eclipse/hawkbit/ddi/rest/resource/DataConversionHelper.java [hunk 9]
```diff
@@ -137,7 +137,8 @@ static DdiControllerBase fromTarget(final Target target, final Action action,
         if (target.isRequestControllerAttributes()) {
             result.add(ControllerLinkBuilder
                     .linkTo(ControllerLinkBuilder.methodOn(DdiRootController.class, tenantAware.getCurrentTenant())
-                            .putConfigData(null, tenantAware.getCurrentTenant(), target.getControllerId()))
+                            .putConfigData(null, tenantAware.getCurrentTenant(), target.getControllerId(),
```
Comment (stefbehl) [inline hawkbit-ddi-resource/src/main/java/org/eclipse/hawkbit/ddi/rest/resource/DataConversionHelper.java]: can be removed once we retrieve the mode from the request body

File: hawkbit-core/src/main/java/org/eclipse/hawkbit/exception/SpServerError.java [hunk 10]
```diff
@@ -17,74 +17,96 @@
     *
     */
     SP_REPO_GENERIC_ERROR("hawkbit.server.error.repo.genericError", "unknown error occured"),
+
     /**
     *
     */
-    SP_REPO_ENTITY_ALRREADY_EXISTS("hawkbit.server.error.repo.entitiyAlreayExists", "The given entity already exists in database"),
+    SP_REPO_ENTITY_ALRREADY_EXISTS("hawkbit.server.error.repo.entitiyAlreayExists",
```
Comment (kaizimmerm) [inline hawkbit-core/src/main/java/org/eclipse/hawkbit/exception/SpServerError.java]: Already

File: hawkbit-ddi-resource/src/main/java/org/eclipse/hawkbit/ddi/rest/resource/DdiRootController.java [hunk 11]
```diff
@@ -427,9 +429,9 @@ private Status handleClosedCase(final DdiActionFeedback feedback, final String c
     @Override
     public ResponseEntity<Void> putConfigData(@Valid @RequestBody final DdiConfigData configData,
             @PathVariable("tenant") final String tenant, @PathVariable("controllerId") final String controllerId) {
-        controllerManagement.updateControllerAttributes(controllerId, configData.getData());
 
-        return ResponseEntity.ok().build();
+        controllerManagement.updateControllerAttributes(controllerId, configData.getData(), getUpdateMode(configData));
+        return ResponseEntity.ok().<Void> build();
```
Comment (kaizimmerm) [inline hawkbit-ddi-resource/src/main/java/org/eclipse/hawkbit/ddi/rest/resource/DdiRootController.java]: I think you can remove the '<Void>'

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/UpdateMode.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+     */
+    REPLACE,
+
+    /**
+     * Removal update strategy
+     */
+    REMOVE;
+
+    public static Optional<UpdateMode> valueOfIgnoreCase(final String name) {
+        return Arrays.stream(values()).filter(mode -> mode.name().equalsIgnoreCase(name)).findFirst();
```
Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/UpdateMode.java]: I think `findAny` is correct here, you do not care about order I guess.

**Pull request merged**